### PR TITLE
Add ops(gridsample, gridsampler2d) | feat(atenlib)

### PR DESCRIPTION
**Pull request by xiaowuhu**

### CODE REVIEW

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 1]
```diff
@@ -869,6 +876,16 @@ def _where_input_wrangler(
         matcher=lambda sample: len(sample.kwargs) == 0 or sample.kwargs.get("p", 0.0) > 0.0,
         reason="dropout is random so the result not match",
     ),
+    skip(
+        "nn.functional.grid_sample",
+        matcher=lambda sample: sample.kwargs.get("mode") == "bicubic",
```
Comment (fatcat-z) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:881]: Current PyTorch exporter tests actually covers this case where mode == 'bicubic', could you please take a look at it?

https://github.com/pytorch/pytorch/blob/a8065cc61f603e61cc166b9a997e2d69e01b7267/test/onnx/test_pytorch_onnx_onnxruntime.py#L12851

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 2]
```diff
@@ -2313,10 +2313,17 @@ def aten_fix(self: TensorType) -> TensorType:
...
+@torch_op("aten::flip")
+def aten_flip(self: TTensor, dims: INT64) -> TTensor:
     """flip(Tensor self, int[] dims) -> Tensor"""
 
-    raise NotImplementedError()
+    shape_dim = op.Shape(dims)
+    neg_1 = op.Constant(value_int=-1)
+    starts = op.Expand(neg_1, shape_dim)  # something like [-1, -1, -1]
+    steps = op.Expand(neg_1, shape_dim)  # something like [-1, -1, -1]
+    ends = starts * 65535  # something like [-65535, -65535, -65535]
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Good point. 65535 seems small. We have _INT64_MAX defined in the file. Possible to define _INT_MIN as well?

[the remark this answers is not in the document]
Comment (xiaowuhu) [reply onnxscript/function_libs/torch_aten/ops/core.py]: define _INT8_MIN= -127, I think it is good enough for tensor dimension

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 1 above)
Comment (xiaowuhu) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:884]: For "bicubic" mode, the result of op.GridSample() is quite different than torch's result. We have no idea what happened. So I have add "FIXME" to mark it.

Comment (xiaowuhu) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:917]: All the 4 case for 'bicubic' mode are failed, so I think it is something different in ORT than Torch. I added a "FIXME" in the skip reason.
```
_ TestOutputConsistencyEagerCPU.test_output_match_opinfo__nn_functional_grid_sample_cpu_float32 (inputs="['Tensor<torch.Size([2, 3, 5, 5]), dtype=torch.float32>', 'Tensor<torch.Size([2, 5, 5, 2]), dtype=torch.float32>']", kwargs="{'mode': 'bicubic', 'padding_mode': 'border', 'align_corners': False}", sample_num=14)
```

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 3]
```diff
@@ -2321,7 +2321,7 @@ def aten_flip(self: TTensor, dims: INT64) -> TTensor:
     neg_1 = op.Constant(value_int=-1)
     starts = op.Expand(neg_1, shape_dim)  # something like [-1, -1, -1]
     steps = op.Expand(neg_1, shape_dim)  # something like [-1, -1, -1]
-    ends = starts * 65535  # something like [-65535, -65535, -65535]
+    ends = starts * _INT8_MIN  # something like [-127, -127, -127]
     result = op.Slice(self, starts, ends, dims, steps)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2325]: Ends in start is defined as `ends (heterogeneous) - Tind: 1-D tensor of ending indices (exclusive) of corresponding axis in axes` and `Tind ` is in ( tensor(int32), tensor(int64) ). https://onnx.ai/onnx/operators/onnx__Slice.html

I think we should just default to use int64 in all cases

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 4]
```diff
@@ -2516,16 +2516,30 @@ def aten_grid_sampler(
     raise NotImplementedError()
 
 
+@torch_op("aten::grid_sampler_2d", trace_only=True)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2519]: Can this be scripted?

Comment (xiaowuhu) [reply onnxscript/function_libs/torch_aten/ops/core.py:2519]: not get your point

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 5]
```diff
@@ -2321,7 +2321,7 @@ def aten_flip(self: TTensor, dims: INT64) -> TTensor:
     neg_1 = op.Constant(value_int=-1)
     starts = op.Expand(neg_1, shape_dim)  # something like [-1, -1, -1]
     steps = op.Expand(neg_1, shape_dim)  # something like [-1, -1, -1]
-    ends = starts * 65535  # something like [-65535, -65535, -65535]
+    ends = starts * _INT64_MIN  # something like [-xxx, -xxx, -xxx]
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2324]: Is this right? Starts is -1. Multiplying it with INT64_MIN will overflow?

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:2324]: Maybe just starts * INT64_MAX

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 6]
```diff
@@ -431,6 +431,26 @@ def aten_glu_jvp(glu: TensorType, x: TensorType, dx: TensorType, dim: int) -> Te
     raise NotImplementedError()
 
 
+@torch_op("aten::grid_sample")
+def aten_grid_sample(
+    input: TTensor,
+    grid: TTensor,
+    mode: str,
+    padding_mode: str,
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:439]: These should be int?

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 5 above)
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2324]:
```suggestion
    ends = op.Expand(_INT64_MIN, shape_dim)  # something like [-xxx, -xxx, -xxx]
```

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 7]
```diff
@@ -2516,16 +2516,30 @@ def aten_grid_sampler(
...
+    grid: TTensor,
     interpolation_mode: int,
     padding_mode: int,
     align_corners: bool,
-) -> TensorType:
+) -> TFloat:
     """grid_sampler_2d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> Tensor"""
 
-    raise NotImplementedError()
+    inter_mode_options = ["bilinear", "nearest", "bicubic"]
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2529]:
```suggestion
    inter_mode_options = ("bilinear", "nearest", "bicubic")
```

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 8]
```diff
@@ -2516,16 +2516,30 @@ def aten_grid_sampler(
...
+    inter_mode_str = inter_mode_options[interpolation_mode]
+
+    padding_mode_options = ["zeros", "border", "reflection"]
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2532]:
```suggestion
    padding_mode_options = ("zeros", "border", "reflection")
```

nit

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 9]
```diff
@@ -811,6 +813,11 @@ def _where_input_wrangler(
         matcher=lambda sample: sample.kwargs.get("rounding_mode") is not None,
         reason="rounding_mode is not yet supported",
     ),
+    skip(
+        "grid_sampler_2d",
+        matcher=lambda sample: sample.args[1] == 2,
+        reason="'bicubic' mode in ORT implemented differently with Torch",
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:819]:
```suggestion
        reason="fixme: 'bicubic' mode in ORT implemented differently with Torch",
```

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 10]
```diff
@@ -909,6 +916,17 @@ def _where_input_wrangler(
...
+    skip(
+        "nn.functional.grid_sample",
+        # FIXME: check why ort's GridSample() return different result than Torch
+        matcher=lambda sample: sample.kwargs.get("mode") == "bicubic",
+        reason="'bicubic' mode in ORT implemented differently with Torch",
+    ),
+    skip(
+        "nn.functional.grid_sample",
+        matcher=lambda sample: len(sample.args[0].shape) != 4,
+        reason="'ORT only support rank(tensor)=4 as input",
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:928]:
```suggestion
        reason="fixme: 'ORT only support rank(tensor)=4 as input",
```

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 11]
```diff
@@ -909,6 +916,17 @@ def _where_input_wrangler(
         matcher=lambda sample: len(sample.kwargs) == 0 or sample.kwargs.get("p", 0.0) > 0.0,
         reason="dropout is random so the result not match",
     ),
+    skip(
+        "nn.functional.grid_sample",
+        # FIXME: check why ort's GridSample() return different result than Torch
+        matcher=lambda sample: sample.kwargs.get("mode") == "bicubic",
+        reason="'bicubic' mode in ORT implemented differently with Torch",
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:923]:
```suggestion
        reason="fixme: 'bicubic' mode in ORT implemented differently with Torch",
```

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 9 above)
Comment (xiaowuhu) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:819]: Torch implemented this using the cubic convolution algorithm with alhpa=−0.75, might be different than ORT. So it maybe by design.

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 11 above)
Comment (xiaowuhu) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:923]: Torch implemented this using the cubic convolution algorithm with alhpa=−0.75, might be different than ORT

**Review by justinchuby — APPROVED** [review]
Thank you!
